---
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: show
    highlight: zenburn
    theme: flatly
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=TRUE, cache.lazy=FALSE, highlight=TRUE, autodep=TRUE,
                      warning=FALSE, error=FALSE,
                      message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

# Overview
This data is RNA-seq data from monocytes and CD4 T cells from healthy control
patients, patients with relapsing-remitting MS (RRMS) and patients with
secondary progessive MS (SPMS). The monocytes and CD4 T cells are taken from the
same sample. Each sample was run twice as technical replicates. The affected
were age matched with the controls, so there are 3 patients that should be
analyzed together if possible in each group. They would like to look at all
pairwise comparisons with the controls, RRMS and SPMS samples within each cell
type. They are not interested in looking across the different cell types.

The data is RNA-seq 3'DGE data-- it appears as if it is SCRB-seq data to me or
something similar to that. The table of counts have been provided, we have the
total counts and the UMI deduped counts; we'll be using the deduped counts
provided by the Broad.

This work is for the Gandhi lab, and Maria Mazzola is the person leading the
project.

# Data cleaning
First we'll read in the count data generated by the Broad and convert the
column names to the well numbers:

```{r read-data}
library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(cowplot)
library(viridis)
counts = read_tsv("../data/SSF-1505_plate_GTAGAGGA.unq.refseq.umi.dat") %>%
  as.data.frame()
rownames(counts) = counts[, 1]
counts[,1] = NULL
cnames = unlist(lapply(strsplit(colnames(counts), split="_"), tail, n=1))
colnames(counts) = cnames
head(counts)
```

We'll also read in the CSV of metadata and place some of the information
into more specific columns. We added columns for `phenotype`, `group`,
`celltype` and `sample` to make it easier to work with the samples later.

```{r read-metadata}
metadata = read_csv("../metadata/sample-info.csv") %>%
  tidyr::separate(samplename, c("sample", "celltype"), sep="_")
metadata$phenotype = substr(metadata$sample, 1, 2)
metadata$group = as.factor(substr(metadata$sample, 3, 3))
metadata$treated = ifelse(metadata$past_treat == "none", "untreated", "treated")
head(metadata)
```

# Sanity checking
Now we'll sanity check the metadata. Samples in the sample group should be
all the same gender and should be similar in age. We'll plot those to make
sure I didn't screw anything up.

```{r group-plots}
ggplot(metadata, aes(group, age, color=gender, size=edss)) +
  geom_jitter() + theme(text=element_text(family='Gill Sans')) +
  scale_color_viridis(discrete=TRUE)
```

Looks good.

The count data has the correct number of columns `r ncol(counts)` and a
reasonable number of rows `r nrow(counts)`. There are no missing values for
the counts or anything like that so that data looks good to go as well.

# Count based quality analysis
Here we calculate a bunch of sumary statistics at the level of the cell.
```{r cell-stats}
cellstats = data.frame(cell=colnames(counts),
                       total_counts=colSums(counts),
                       genes_detected=colSums(counts > 0),
                       robust_genes_detected=colSums(counts > 10)) %>%
  left_join(metadata, by="cell")
```

## Total counts
Overall the total counts per cell has a pretty even distribution, hovering
around 0.5 - 1 million counts per cell. Groups 7 and 8 seem to have a
consistently lower counts per cell than the rest of the samples. Usually I colored the
bars by `rin` number, `age`, `gender` and `viable_pct` (not shown) but none of
those seemed to correspond to the lower number of counts.

```{r total-counts}
ggplot(cellstats, aes(cell, total_counts)) +
  geom_bar(stat='identity') +
  facet_wrap(~group, scales='free_x') +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size=8),
        text=element_text(family='Gill Sans')) +
  xlab("") +
  ylab("total counts")
```

## Genes detected
The plots of genes detected looks pretty good, a nice even amount for all of the
samples. The number of genes detected is high, around half of the queried genes
per cell, which is good. Groups 7 and 8 tend to have a slightly lower number of
genes detected than the other cells, which we expected due to having a lower
number of counts.
```{r genes-detected}
ggplot(cellstats, aes(cell, genes_detected)) +
  geom_bar(stat='identity') +
  facet_wrap(~group, scales='free_x') +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size=8),
        text=element_text(family='Gill Sans')) +
  xlab("") +
  ylab("genes detected")
```

## Library complexity
Here we plot genes detected vs the total number of counts. We see a clear
relationship between the number of genes detected and the total number of
counts, suggesting that we could detect even more genes if we sequenced these
samples more deeply. This plot is mainly to detect if there are groups
of failed cells that don't seem to follow along the polynomial line we fit.

```{r library-complexity}
ggplot(cellstats, aes(total_counts, genes_detected)) +
  geom_point() +
  geom_smooth(method="lm", formula=y~poly(x, 2), size=1) +
  theme(text=element_text(family='Gill Sans')) +
  xlab("total counts") +
  ylab("genes detected")
```

# PCA {.tabset}
Here we run principal component analysis (PCA) on the count data. Before performing
PCA we center and scale the data and correct for the mean-variance relationship
in RNA-seq data by performing a variance stabilizing transformation on the data.
This ensures that low expressed genes with high variance do not dominate the
PCA plot.

After variance stabilizing the data we take the top 500 most variable genes and
use them to perform PCA.

```{r pca-plots}
library(DESeq2)
design = ~group+celltype+phenotype+phenotype:celltype
dds = DESeqDataSetFromMatrix(countData=counts, colData=metadata, design=design)
vst = varianceStabilizingTransformation(dds)
pca_loadings = function(object, ntop=500) {
  rv <- matrixStats::rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop,
      length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  names(percentVar) = colnames(pca$x)
  pca$percentVar = percentVar
  return(pca)}
pc = pca_loadings(vst)
comps = data.frame(pc$x)
comps$Name = rownames(comps)
library(dplyr)
comps = comps %>% left_join(metadata, by=c("Name"="cell"))
```

```{r pca-plotter}
pca_plot = function(comps, nc1, nc2, colorby) {
  library(cowplot)
  library(viridis)
   c1str = paste0("PC", nc1)
   c2str = paste0("PC", nc2)
  ggplot(comps, aes_string(c1str, c2str, color=colorby)) +
    geom_point() +
    xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), "% variance")) +
    ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 100), "% variance"))
}
```

## PC1 vs. PC2
We can see a great separation by cell type along PC1:
```{r PC1-PC2-celltype}
pca_plot(comps, 1, 2, "celltype")
```

## PC2 vs. PC3
We see plotting PC2 vs PC3 a more subtle difference between phenotypes. RR and MS
look mixed but there seems to be a separation between the control and the
affected samples.

```{r PC2-PC3-phenotype}
pca_plot(comps, 2, 3, "phenotype")
```

# PCA, cell type separation
Part of the reason for the poor higher component separation is we are using the
500 genes that are dominated by cell type. If we first separate out the cell
types and then do the PCA, we might see a better separation based on phenotype.

```{r separate-types}
monometa = subset(metadata, celltype == "mono")
monocounts = counts[, monometa$cell]
tcellmeta = subset(metadata, celltype == "Tcells")
tcellcounts = counts[, tcellmeta$cell]
design = ~group+phenotype
monodds = DESeqDataSetFromMatrix(countData=monocounts, colData=monometa,
                                 design=design)
tcelldds = DESeqDataSetFromMatrix(countData=tcellcounts, colData=tcellmeta,
                                  design=design)
monovst = varianceStabilizingTransformation(monodds)
tcellvst = varianceStabilizingTransformation(tcelldds)
monopc = pca_loadings(monovst)
tcellpc = pca_loadings(tcellvst)
monocomps = data.frame(monopc$x)
tcellcomps = data.frame(tcellpc$x)
monocomps$cell = rownames(monocomps)
tcellcomps$cell = rownames(tcellcomps)
monocomps = monocomps %>% left_join(metadata, by=c("cell"="cell"))
tcellcomps = tcellcomps %>% left_join(metadata, by=c("cell"="cell"))
```

## T-cell {.tabset}

### PC1-PC2
We can see a separation along component 2 with which `gender` the samples are from:

```{r PC1-PC2-tcell}
pca_plot(tcellcomps, 1, 2, "gender")
pca_plot(tcellcomps, 1, 2, "phenotype")
pca_plot(tcellcomps, 1, 2, "cur_treat")
pca_plot(tcellcomps, 1, 2, "past_treat")
pca_plot(tcellcomps, 1, 2, "duration") + scale_color_viridis()
```

But component 1 is a little bit of a mystery, it does kind of separate by
phenotype, all of the HC samples are clustered on the left and the right is more
mixed MS/RR.

### PC3-PC4
```{r PC3-PC4-tcell}
pca_plot(tcellcomps, 3, 4, "gender")
pca_plot(tcellcomps, 3, 4, "phenotype")
pca_plot(tcellcomps, 3, 4, "cur_treat")
pca_plot(tcellcomps, 3, 4, "past_treat")
pca_plot(tcellcomps, 3, 4, "duration") + scale_color_viridis()
```

## Monocyte {.tabset}
We don't see the same gender based separation in the monocytes that we saw
with the Tcells.

### PC1-PC2
There isn't a separation based on gender in PC1-PC2, it isn't clear what
is separating these cells because they also don't separate by phenotype.
They don't seem to separate very well on current treatment or by

```{r PC1-PC2-mono}
pca_plot(monocomps, 1, 2, "gender")
pca_plot(monocomps, 1, 2, "phenotype")
pca_plot(monocomps, 1, 2, "cur_treat")
pca_plot(monocomps, 1, 2, "past_treat")
pca_plot(monocomps, 1, 2, "duration") + scale_color_viridis()
```

### PC3-PC4
Samples do separate along PC3 and PC4 by gender. So for T-cells there seems to
be a larger gender effect than the monocytes.

```{r PC3-PC4-mono}
pca_plot(monocomps, 3, 4, "gender")
pca_plot(monocomps, 3, 4, "phenotype")
pca_plot(monocomps, 3, 4, "cur_treat")
pca_plot(monocomps, 3, 4, "past_treat")
pca_plot(monocomps, 3, 4, "duration") + scale_color_viridis()
```

# Filtering and combining
The cells all look good so we don't need to do any filtering of the individual
cells. We'll drop all genes that don't have any counts in any cells, and then
combine the counts for all of the technical replicate cells by adding them
together before doing differential expression.

```{r filter-and-combine}
monomelt = monocounts %>% add_rownames() %>%
  tidyr::gather(cell, count, -rowname) %>%
  left_join(monometa, by="cell") %>%
  select(rowname, sample, count) %>%
  group_by(rowname, sample) %>%
  summarise(total=sum(count)) %>%
  tidyr::spread(sample, total) %>% as.data.frame()
rownames(monomelt) = monomelt$rowname
monomelt$rowname = NULL
monomelt = monomelt[rowSums(monomelt) > 0,]
monometa = monometa %>% group_by(sample) %>%
  filter(row_number() == 1) %>%
  select(-cell)
monometa = as.data.frame(monometa)
rownames(monometa) = monometa$sample
monomelt = monomelt[, rownames(monometa)]

tcellmelt = tcellcounts %>% add_rownames() %>%
  tidyr::gather(cell, count, -rowname) %>%
  left_join(tcellmeta, by="cell") %>%
  select(rowname, sample, count) %>%
  group_by(rowname, sample) %>%
  summarise(total=sum(count)) %>%
  tidyr::spread(sample, total) %>% as.data.frame()
rownames(tcellmelt) = tcellmelt$rowname
tcellmelt$rowname = NULL
tcellmelt = tcellmelt[rowSums(tcellmelt) > 0,]
tcellmeta = tcellmeta %>% group_by(sample) %>%
  filter(row_number() == 1) %>%
  select(-cell)
tcellmeta = as.data.frame(tcellmeta)
rownames(tcellmeta) = tcellmeta$sample
tcellmelt = tcellmelt[, rownames(tcellmeta)]
```

# Differential expression
We will fit the model to the two different celltypes separately. We fit
a model of the form: `~group+phenotype` where we compare only the cells
from the same group to each other and look for differences in phenotype.

## Monocytes
The dispersion plot looks awesome, in this plot the black dots are the
estimates of the dispersion if you just considered each gene by itself,
and the red line is the best fit of the dispersion for genes of a
given expression value. The blue dots are the final value for the dispersions,
which for outlier spots are pulled back towards the fitted line.

```{r mono-deseq2}
design = ~group+phenotype
monodds = DESeqDataSetFromMatrix(countData=monomelt,
                                 colData=monometa, design=design)
monodds = estimateSizeFactors(monodds)
monodds = DESeq(monodds)
plotDispEsts(monodds)
```

### RR vs HC
```{r mono-hc-rr}
monorr_vs_hc = results(monodds, contrast=c("phenotype", "RR", "HC"))
plotMA(monorr_vs_hc)
monorr_vs_hc = monorr_vs_hc[order(monorr_vs_hc$padj),]
monorr_vs_hc = data.frame(cbind(data.frame(gene=rownames(monorr_vs_hc)), monorr_vs_hc))
monorr_vs_hc_sig = subset(monorr_vs_hc, padj < 0.1)
write.table(monorr_vs_hc, file="mono-rr-vs-hc.csv", row.names=FALSE,
            col.names=TRUE, quote=FALSE, sep=",")
```

We flag `r nrow(monorr_vs_hc_sig)` genes different between `RR` and `HC`
samples in the monocytes, using a FDR cutoff of 0.1. Of those
`r nrow(subset(monorr_vs_hc_sig, log2FoldChange > 0))`
are up in the `RR` samples and `r nrow(subset(monorr_vs_hc_sig, log2FoldChange < 0))`
are down in the `RR` samples.

### SP vs HC
```{r mono-vs-sp}
monosp_vs_hc = results(monodds, contrast=c("phenotype", "SP", "HC"))
plotMA(monosp_vs_hc)
monosp_vs_hc = monosp_vs_hc[order(monosp_vs_hc$padj),]
monosp_vs_hc = data.frame(cbind(data.frame(gene=rownames(monosp_vs_hc)), monosp_vs_hc))
monosp_vs_hc_sig = subset(monosp_vs_hc, padj < 0.1)
write.table(monosp_vs_hc, file="mono-sp-vs-hc.csv", row.names=FALSE,
            col.names=TRUE, quote=FALSE, sep=",")
```

We flag `r nrow(monosp_vs_hc_sig)` genes different between `SP` and `HC`
samples in the monocytes, using a FDR cutoff of 0.1. Of those
`r nrow(subset(monosp_vs_hc_sig, log2FoldChange > 0))`
are up in the `SP` samples and `r nrow(subset(monosp_vs_hc_sig, log2FoldChange < 0))`
are down in the `SP` samples.

### SP vs RR
```{r mono-sp-rr}
monosp_vs_rr = results(monodds, contrast=c("phenotype", "SP", "RR"))
plotMA(monosp_vs_rr)
monosp_vs_rr = monosp_vs_rr[order(monosp_vs_rr$padj),]
monosp_vs_rr = data.frame(cbind(data.frame(gene=rownames(monosp_vs_rr)), monosp_vs_rr))
monosp_vs_rr_sig = subset(monosp_vs_rr, padj < 0.1)
write.table(monosp_vs_rr, file="mono-sp-vs-rr.csv", row.names=FALSE,
            col.names=TRUE, quote=FALSE, sep=",")
```

We flag `r nrow(monosp_vs_rr_sig)` genes different between `SP` and `RR`
samples in the monocytes, using a FDR cutoff of 0.1. Of those
`r nrow(subset(monosp_vs_rr_sig, log2FoldChange > 0))`
are up in the `SP` samples and `r nrow(subset(monosp_vs_rr_sig, log2FoldChange < 0))`
are down in the `SP` samples.


## T-cells
```{r tcells-deseq2}
design = ~group+phenotype
tcelldds = DESeqDataSetFromMatrix(countData=tcellmelt,
                                 colData=tcellmeta, design=design)
tcelldds = estimateSizeFactors(tcelldds)
tcelldds = DESeq(tcelldds)
plotDispEsts(tcelldds)
```

### RR vs HC
```{r tcells-hc-rr}
tcellrr_vs_hc = results(tcelldds, contrast=c("phenotype", "RR", "HC"))
plotMA(tcellrr_vs_hc)
tcellrr_vs_hc = tcellrr_vs_hc[order(tcellrr_vs_hc$padj),]
tcellrr_vs_hc = data.frame(cbind(data.frame(gene=rownames(tcellrr_vs_hc)), tcellrr_vs_hc))
tcellrr_vs_hc_sig = subset(tcellrr_vs_hc, padj < 0.1)
write.table(tcellrr_vs_hc, file="tcell-rr-vs-hc.csv", row.names=FALSE,
            col.names=TRUE, quote=FALSE, sep=",")
```

We flag `r nrow(tcellrr_vs_hc_sig)` genes different between `RR` and `HC`
samples in the T-cells, using a FDR cutoff of 0.1. Of those
`r nrow(subset(tcellrr_vs_hc_sig, log2FoldChange > 0))`
are up in the `RR` samples and `r nrow(subset(tcellrr_vs_hc_sig, log2FoldChange < 0))`
are down in the `RR` samples.

### SP vs HC
```{r tcells-sp-hc}
tcellsp_vs_hc = results(tcelldds, contrast=c("phenotype", "SP", "HC"))
plotMA(tcellsp_vs_hc)
tcellsp_vs_hc = tcellsp_vs_hc[order(tcellsp_vs_hc$padj),]
tcellsp_vs_hc = data.frame(cbind(data.frame(gene=rownames(tcellsp_vs_hc)), tcellsp_vs_hc))
tcellsp_vs_hc_sig = subset(tcellsp_vs_hc, padj < 0.1)
write.table(tcellsp_vs_hc, file="tcell-sp-vs-hc.csv", row.names=FALSE,
            col.names=TRUE, quote=FALSE, sep=",")
```

We flag `r nrow(tcellsp_vs_hc_sig)` genes different between `sp` and `HC`
samples in the T-cells, using a FDR cutoff of 0.1. Of those
`r nrow(subset(tcellsp_vs_hc_sig, log2FoldChange > 0))`
are up in the `SP` samples and `r nrow(subset(tcellsp_vs_hc_sig, log2FoldChange < 0))`
are down in the `SP` samples.

### SP vs RR
```{r tcells-sp-rr}
tcellsp_vs_rr = results(tcelldds, contrast=c("phenotype", "SP", "RR"))
plotMA(tcellsp_vs_rr)
tcellsp_vs_rr = tcellsp_vs_rr[order(tcellsp_vs_rr$padj),]
tcellsp_vs_rr = data.frame(cbind(data.frame(gene=rownames(tcellsp_vs_rr)), tcellsp_vs_rr))
tcellsp_vs_rr_sig = subset(tcellsp_vs_rr, padj < 0.1)
write.table(tcellsp_vs_rr, file="tcell-sp-vs-rr.csv", row.names=FALSE,
            col.names=TRUE, quote=FALSE, sep=",")
```

We flag `r nrow(tcellsp_vs_rr_sig)` genes different between `sp` and `RR`
samples in the T-cells, using a FDR cutoff of 0.1. Of those
`r nrow(subset(tcellsp_vs_rr_sig, log2FoldChange > 0))`
are up in the `SP` samples and `r nrow(subset(tcellsp_vs_rr_sig, log2FoldChange < 0))`
are down in the `SP` samples.

## MS vs control
### MS vs control - monocytes
```{r ms-vs-control-monocytes}
monometa$affected = ifelse(monometa$phenotype == "HC", "unaffected", "affected")
design = ~group+affected
monoaffecteddds = DESeqDataSetFromMatrix(countData=monomelt,
                                 colData=monometa, design=design)
monoaffecteddds = estimateSizeFactors(monoaffecteddds)
monoaffecteddds = DESeq(monoaffecteddds)
monoaffected = results(monoaffecteddds, contrast=c("affected", "affected", "unaffected"))
plotMA(monoaffected)
monoaffected = monoaffected[order(monoaffected$padj),]
monoaffected = data.frame(cbind(data.frame(gene=rownames(monoaffected)), monoaffected))
monoaffected_sig = subset(monoaffected, padj < 0.1)
write.table(monoaffected, file="monoaffected.csv", row.names=FALSE,
            col.names=TRUE, quote=FALSE, sep=",")
```

We flag `r nrow(monoaffected_sig)` genes different between `affected` and
`unaffected` monocytes, using a FDR cutoff of 0.1. Of those
`r nrow(subset(monoaffected_sig, log2FoldChange > 0))` are up in the `affected`
samples and `r nrow(subset(monoaffected_sig, log2FoldChange < 0))` are down in
the `affected` samples.

### MS vs control - Tcells
```{r ms-vs-control-tcells}
tcellmeta$affected = ifelse(tcellmeta$phenotype == "HC", "unaffected", "affected")
design = ~group+affected
tcellaffecteddds = DESeqDataSetFromMatrix(countData=tcellmelt,
                                 colData=tcellmeta, design=design)
tcellaffecteddds = estimateSizeFactors(tcellaffecteddds)
tcellaffecteddds = DESeq(tcellaffecteddds)
tcellaffected = results(tcellaffecteddds, contrast=c("affected", "affected", "unaffected"))
plotMA(tcellaffected)
tcellaffected = tcellaffected[order(tcellaffected$padj),]
tcellaffected = data.frame(cbind(data.frame(gene=rownames(tcellaffected)), tcellaffected))
tcellaffected_sig = subset(tcellaffected, padj < 0.1)
write.table(tcellaffected, file="tcellaffected.csv", row.names=FALSE,
            col.names=TRUE, quote=FALSE, sep=",")
```

We flag `r nrow(tcellaffected_sig)` genes different between `affected` and
`unaffected` tcells, using a FDR cutoff of 0.1. Of those
`r nrow(subset(tcellaffected_sig, log2FoldChange > 0))` are up in the `affected`
samples and `r nrow(subset(tcellaffected_sig, log2FoldChange < 0))` are down in
the `affected` samples.

# Pathway analysis
# Summary
# Output tables

[Monocyte RR vs HC](mono-rr-vs-hc.csv)

[Monocyte SP vs HC](mono-sp-vs-hc.csv)

[Monocyte SP vs RR](mono-sp-vs-rr.csv)

[Monocyte affected vs unaffected](monoaffected.csv)

[T-Cell RR vs HC](tcell-rr-vs-hc.csv)

[T-Cell SP vs HC](tcell-sp-vs-hc.csv)

[T-Cell SP vs RR](tcell-sp-vs-rr.csv)

[T-cell affected vs unaffected](tcellaffected.csv)

# Normalized counts

```{r write-normalized-counts}
nmonocounts = counts(monodds, normalized=TRUE)
nmonocounts = cbind(data.frame(gene=rownames(nmonocounts)), nmonocounts)
write.table(nmonocounts, file="monocounts_normalized.csv", row.names=FALSE,
            col.names=TRUE, sep=",", quote=FALSE)
ntcellcounts = counts(tcelldds, normalized=TRUE)
ntcellcounts = cbind(data.frame(gene=rownames(ntcellcounts)), ntcellcounts)
write.table(ntcellcounts, file="tcellcounts_normalized.csv", row.names=FALSE,
            col.names=TRUE, sep=",", quote=FALSE)
```

[Normalized counts for monocytes](monocounts_normalized.csv)

[Normalized counts for T cells](tcellcounts_normalized.csv)
